### ISSUE TITLE
Feature request: allow rotary cos/sin to vary per head

### ISSUE THREAD

**Issue by crowsonkb** [Dao-AILab/flash-attention#578]
I have noticed that rotary does better on my model, especially for small heads which have few rotary frequencies, when I apply a different set of frequencies per head (it is able to use the difference between the heads to extract finer position information). I did this using my own pure PyTorch apply_rotary_emb(), but I currently use apply_rotary_emb_qkv_() for speed and memory savings and that does not support a sin/cos that is different for each head. Could this be added?

Thank you,
Katherine Crowson

**Comment by lucidrains** [comment]
@crowsonkb that's really interesting Katherine! did you share these experiments somewhere?

**Comment by lucidrains** [comment]
@crowsonkb hey, i actually have a need to use your idea (working with limited head dimension). can i cite you somewhere, if you tweeted about this or something?

**Comment by samgd** [comment]
@lucidrains any tips or findings?

for our smallest models with small head dim we think the model might be limited somewhat by the position embedding so this could be a win

have started a few runs with varying theta schemes/custom freqs and will share if any work!

**Comment by lucidrains** [comment]
@samgd i haven't had the time for experiments yet, but will get around to it! was going to try bottlenecking head dimension to 16 and if a signal is seen, use [learned frequencies](https://github.com/lucidrains/rotary-embedding-torch/blob/main/rotary_embedding_torch/rotary_embedding_torch.py#L73) and see what the network converges on

very interested in your results!

**Comment by samgd** [comment]
👍 thanks

head dim is 16 here so we will see!

learned already underway too 😊

(fyi I think your rotary lib learned freqs might be broken atm as I had to clear the cache before each forward pass to prevent a backprop error, not had time to look into a real solution/make an issue yet!)

**Comment by lucidrains** [comment]
@samgd oops! put in a quick fix 🙏

**Comment by lucidrains** [comment]
@samgd did you see anything Sam? i think i may implement it today for image self attention, where the dimensions is halved due to the axial positions

**Comment by samgd** [comment]
:wave: yeah so i ran a bunch of variants on short (~0.5 day) runs and they all ended up basically identical, with baseline default parameter full rotation coming out ever so slightly on top

variants were:

- learned per-layer frequencies, shared by all heads within a layer
- ^ with no weight decay on frequencies
- ^ but allowing the frequencies to vary per head
- ^ but randomly initialised rather than being initialised from baseline frequencies
- fixed but varying frequencies per-head, per-head. i.e. change theta parameter to be linspace from 100 to 20000 for each head
- ^ but a slightly different distribution

this was on an unusual task, although one you're familiar with :grin: nanopore hac-sized transformer basecaller

if you do find anything i am interested, negative results from me but still suspect there might be something here

**Comment by samgd** [comment]
lots to do still but interesting results internally so we'll see!

yeah we met ice cream :grin:
